### PR TITLE
async boxers / unboxers

### PR DESCRIPTION
**Pull request by mixmix**
WIP:

:fire: This is a breaking change

## Motivation
I need to be able to do async lookups to look for appropriate keys for private-groups. 
This requires changing how boxers / unboxers work from sync > async.

---

## Questions

what in the ecosystem is currently using these APIs `box` / `unbox`, `addBoxer` / `addUnboxer`?

## TODO

if we're cutting a major release I intend to extract the box1 boxers/ unboxers so that they are optionally added as opposed to always on.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
...
 
 ### db.addUnboxer({key: unboxKey, value: unboxValue})
 
 add an unboxer object, any encrypted message is passed to the unboxer object to
 test if it can be unboxed (decrypted)
 
 where
-- `unboxKey(ciphertext) => msgKey` is a function which tries to extract the message key from the encrypted content (`ciphertext`)
-- `unboxValue(ciphertext, msgKey) => plaintext` is a function which takes a message key and uses it to try to extract the message content from the `ciphertext`
+- `unboxKey(msg.value.content, msg.value, cb)`
```
Comment (christianbundy) [inline README.md]: Nit: Might be nice to be consistent here -- the `box()` signature references `content` and `recps` but this one references `msg.value.content` and `msg.value`.

File: minimal.js [hunk 2]
```diff
@@ -95,7 +95,15 @@ module.exports = function (dirname, keys, opts) {
 
   var log = OffsetLog(path.join(dirname, 'log.offset'), { blockSize: 1024 * 16, codec })
 
-  const unboxerMap = (msg, cb) => cb(null, db.unbox(msg))
+  const unboxerMap = (msg, cb) => {
+    db.unbox(msg, null, (err, unboxed) => {
+      if (err) {
+        console.error(err)
+        return cb(null, msg)
+      }
```
Comment (mixmix) [inline minimal.js]: I'd like input thinking about how to handle error message with boxing / unboxing.

Comment (mixmix) [reply minimal.js]: This here allows errors to bubble up through callbacks but then pops a log out and "returns" an unecrypted message.

File: README.md [hunk 3]
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
...
+  - is a function which tries to extract the message key from the encrypted content (`ciphertext`).
+  - is expected to callback `cb(err, msgKey)` which `msgKey` is the key for the message
+- `unboxValue(msg.value.content, msgKey, msg.value, cb)`
```
Comment (christianbundy) [inline README.md]: I'm sure there's a good reason why, but I don't understand:  could this just be `unboxValue(msg, cb)`? Seems like we've destructured `msg` into its parts and are passing `content` twice (in `msg.value.content` and `msg.value`).

File: minimal.js [hunk 4]
```diff
@@ -226,38 +239,68 @@ module.exports = function (dirname, keys, opts) {
...
+        break
+
+      case 'object':
+        if (typeof unboxer.key !== 'function') throw new Error('invalid unboxer')
+        if (typeof unboxer.value !== 'function') throw new Error('invalid unboxer')
+        unboxers.push(unboxer)
+        break
+
+      default: throw new Error('invalid unboxer')
+    }
```
Comment (mixmix) [inline minimal.js]: I decided to guard against weird stuff being poked in here

File: minimal.js [hunk 5]
```diff
@@ -226,38 +239,68 @@ module.exports = function (dirname, keys, opts) {
...
+    unboxer: {
+      key: function (ciphertext, value, cb) { 
+        if (!ciphertext.endsWith('.box')) return cb(null, null)
+        // TODO move this inside of ssb-keys
+
+        cb(null, ssbKeys.unboxKey(ciphertext, keys))
+      },
+      value: function (ciphertext, msgKey, value, cb) {
+        cb(null, ssbKeys.unboxBody(ciphertext, msgKey))
+      }
```
Comment (mixmix) [inline minimal.js]: I'd like a review on the signatures of these boxers / unboxers.
What adding a `cb` arg to the end surfaced is that the current signature has args which aren't being used (i.e. the `value = msg.value`).

I am tempted to refactor these functions to have signatures like : 

```js
function unboxValue (ciphertext, msgKey, context, cb) {
  // ...
}
```
where `context = { msg, ...futureThings }`

I'd also change the `unboxer` signature to `unboxer = { unboxKey: Function, unboxValue: Function }`

File: create.js [hunk 6]
```diff
@@ -83,19 +83,18 @@ module.exports = function create (path, opts, keys) {
...
 
-        let result
-
-        if (isPrivate) {
-          result = data.value
-        } else {
-          result = u.originalValue(data.value)
+            if (meta) cb(null, { key, value: result.value, timestamp: result.timestamp })
+            else cb(null, result.value)
+          })
```
Comment (mixmix) [inline create.js]: I honestly had a hard time reading the original code because the logical paths through it seem woven and unclear to me.

I _think_ what I've written does the same thing. Certainly the tests present pass

Comment (christianbundy) [inline create.js]: Cool! Looks like we're only unboxing when we need to, which is great. Have you ran benchmarks on this code? It's important that it doesn't slow anything down, because this is a very hot path, but I have a hunch that this might even speed up methods like `createHistoryStream()`.

File: minimal.js [hunk 7]
```diff
@@ -226,38 +239,68 @@ module.exports = function (dirname, keys, opts) {
...
+  db.box = function (content, state, cb) { // state not used
+    var recps = content.recps
+    if (!recps) return cb(null, content)
 
-    if (typeof recps === 'string') recps = value.content.recps = [recps]
-    if (!Array.isArray(recps)) throw new Error('private message field "recps" expects an Array of recipients')
-    if (recps.length === 0) throw new Error('private message field "recps" requires at least one recipient')
+    if (typeof recps === 'string') recps = content.recps = [recps]
+    if (!Array.isArray(recps)) return cb(new Error('private message field "recps" expects an Array of recipients'))
+    if (recps.length === 0) return cb(new Error('private message field "recps" requires at least one recipient'))
```
Comment (christianbundy) [inline minimal.js]: Do you think it would be wise to move the `if (!recps.every(isFeed)) return cb(null, null)` code here? Seems like that would be valid for any boxer, not just the default `box1` boxer, but I'm not confident about that.

File: README.md (hunk 1 above)
Comment (mixmix) [reply README.md]: true, if you think reading it like `msg.value.content` etc helps then I think making it consistent is a good idea

File: README.md (hunk 3 above)
Comment (mixmix) [reply README.md]: I'm going to copy this comment into a comment I made about this.
TL;DR I agree

File: minimal.js (hunk 5 above)
Comment (mixmix) [reply minimal.js]: @christianbundy said : 

> I'm sure there's a good reason why, but I don't understand:  could this just be `unboxValue(msg, cb)`? Seems like we've destructured `msg` into its parts and are passing `content` twice (in `msg.value.content` and `msg.value`).

File: create.js (hunk 6 above)
Comment (mixmix) [reply create.js]: From call

use https://github.com/ssbc/bench-ssb

run 1-6
---

run these benches before + after on my system

File: minimal.js (hunk 7 above)
Comment (mixmix) [reply minimal.js]: No this is going to be different with different boxers. The new private groups for example is gonna let you do

```
  recps: [ @<base64 feedId>.sha256, %<base64 groupId>.cloaked ]
```

File: create.js (hunk 6 above)
Comment (mixmix) [reply create.js]: ## Process

- clone down [`bench-ssb`](https://github.com/ssbc/bench-ssb), npm linking `ssb-db`
- ran `run-all.sh` ... saw output turn up, but bench didn't get beyond test 9 for some reason (so I quit out)
- changed branch of `ssb-db` (back to master)
- changed `run-all.sh` to 
  - skip generate (so using same output.json as first test)
  - append `>>` results, instead of overwrite `>` (so can easily compare)
  - skipped tests `>= 9`

## Results

here's a sampling of results from tests 1-8
```
// 04-out.txt
minimal 100001 1 7197.941409342834 13.893 // async-box branch
minimal 100001 1 6958.527590285992 14.371 // master branch
```
```
07-out.txt
clock-dump 99001 9001 14375.054450413823 6.887  // async-box branch
clock-dump 99001 9001 14629.968967045957 6.767  // master branch
```

```
// 08-out.txt
sbot-read 100002 2 10718.327974276528 9.33  // async-box branch
sbot-read 100002 2 9706.105017955935 10.303 // master branch
```

**Comment by mixmix** [comment]
## Update

reviewed this with @dominictarr this afternoon. He's thought of some good changes - TL;DR the methods won't be async but we are extending the API.
Changes to follow

File: minimal.js (hunk 5 above)
Comment (mixmix) [reply minimal.js]: reviewed and pruned with Dominic

**Comment by mixmix** [comment]
I think this is ready for a review if anyone wants to take a look.

**Comment by christianbundy** [comment]
Is it pretty straightforward, or are there interesting changes that you made since we last discussed? I can probably take a look at this tomorrow sometime.

**Comment by mixmix** [comment]
hmm, changes: 
- it's now sync again
  - dominic doesn't want to tempt people to do I/O on box / unbox
  - we figure a lot of keys state is small enough it can / should be in memory
  - we added an async `init` option the `addUnboxer`. This is where people can set state before things kick off
- refactoring
  - pulled out the auto box / unbox into `autobox.js`. way more chill IMO
  - wrote `AsyncJobQueue` to manage trickyness to knowing when things are ready or should wait. (with tests). collects logic nicely IMO
- removed `db.box` from the API exposed by `ssb-db`
  - this is now the only breaking change

File: test/util.js [hunk 8]
```diff
@@ -1,12 +1,92 @@
...
-  opts = opts || {}
-  var dir = require('path').join(require('osenv').tmpdir(), name)
-  if (opts.temp !== false) {
-    require('rimraf').sync(dir)
-    require('mkdirp').sync(dir)
-  }
-  opts.keys = opts.keys || ssbKeys.generate()
-  return require('../create')(dir, opts, opts.keys)
-}
+'use strict'
```
Comment (mixmix) [inline test/util.js]: this file now tests `/util.js` and the original file has been moved to `test/create-ssb.js`

File: test/box-unbox.js [hunk 9]
```diff
@@ -0,0 +1,286 @@
+'use strict'
```
Comment (mixmix) [inline test/box-unbox.js]: this file used to be called `end-to-end.js`. I thought the name was misleading.
It's been renamed, and it's ~85% the same. I had to change some parts, and refactored a little as I went to be able to parse what the tests were doing

Importantly I've added tests around : 
- `addBoxer`
- `addUnboxer` (both signatures, and new async `init`)

File: README.md [hunk 10]
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
...
+### db.addBoxer(boxer)
 
-add a box, which will be added to the list of boxers which will try to
+add a `boxer`, which will be added to the list of boxers which will try to
 automatically box (encrypt) the message `content` if the appropriate
 `content.recps` is provided.
 
-`box` is a function of signature `box(content, recps) => (String|null)`
-which is expected to either box the the message content and return a ciphertext String, or return null if it unable to.
+`box` is a function of signature `box(content, recps) => ciphertext` which is expected either:
```
Comment (arj03) [inline README.md]: Think this should be `boxer`?

File: README.md [hunk 11]
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
...
 test if it can be unboxed (decrypted)
 
 where
-- `unboxKey(ciphertext) => msgKey` is a function which tries to extract the message key from the encrypted content (`ciphertext`)
-- `unboxValue(ciphertext, msgKey) => plaintext` is a function which takes a message key and uses it to try to extract the message content from the `ciphertext`
+- `unboxKey(msg.value.content, msg.value) => msgKey`
+  - is a function which tries to extract the message key from the encrypted content (`ciphertext`).
+  - is expected to return `msgKey` which is the key for the message
+- `unboxValue(msg.value.content, msgKey) => plainContent`
+  - is a function which takes a `msgKey` and uses it to try to extract the `plainContent` from the `ciphertext- `initBoxer(done)`
```
Comment (arj03) [inline README.md]: indention?

File: minimal.js [hunk 12]
```diff
@@ -10,81 +10,7 @@ var ssbKeys = require('ssb-keys')
...
-
-## append (msg, cb)
-
-write a message, callback once it's definitely written.
-*/
-
-function isString (s) {
-  return typeof s === 'string'
-}
+var { box, unbox } = require('./autobox')
```
Comment (arj03) [inline minimal.js]: :heart:

File: util.js [hunk 13]
```diff
@@ -159,3 +159,50 @@ exports.metaBackup = (msgValue, property) => {
 
   return msgValue.meta
 }
+
+exports.AsyncJobQueue = class AsyncJobQueue {
```
Comment (arj03) [inline util.js]: Someone must have made a module for this already?

**Comment by arj03** [comment]
Made a rather quick review. I havn't run any of this code. Was only reading the changes.

File: util.js (hunk 13 above)
Comment (mixmix) [reply util.js]: yes, this started out as a minimal refactor and as I added tests got more complex.
Given it's working and working (I think) exactly like the code which was previously there, I'm kinda keen to leave this as is.

Feel free anyone to drop an alternative in and see if it passes the same tests!

**Comment by staltz** [comment]
I like to review files in a meaningful order, not the lexicographic order, do you think this order is meaningful (sorted from important to auxiliary):

```
 README.md               |  26 +++++--
 test/box-unbox.js       | 286 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 test/create-ssb.js      |  12 ++++
 test/close.js           |  22 +++---
 test/util.js            | 104 +++++++++++++++++++++++----
 test/end-to-end.js      | 170 --------------------------------------------
 autobox.js              |  84 ++++++++++++++++++++++
 minimal.js              | 257 +++++++++++++++++++++++++++----------------------------------------
 create.js               |  17 ++---
 index.js                |   1 -
 util.js                 |  47 +++++++++++++
 test/add.js             |   2 +-
 test/feed.js            |   2 +-
 test/history.js         |   2 +-
 test/latest.js          |   2 +-
 test/links.js           |   2 +-
 test/linktypes.js       |   2 +-
 test/log.js             |   2 +-
 test/mesages-by-type.js |   2 +-
 test/msg-encoding.js    |   2 +-
 test/validation.js      |   2 +-
 test/write-stream.js    |   2 +-
```

File: README.md (hunk 11 above)
Comment (mixmix) [reply README.md]: I have applied more teeth  

_indention !== indentation_ :heart:

**Comment by mixmix** [comment]
@staltz I'll split out a different PR which just moves and renames somethings which will reduce noise here

**Comment by mixmix** [comment]
There's an intermitent fail in `test/close.js`. Not my fault I am 99% sure.

@staltz this should be easier to review.
The guts of the changes are in `minimal.js`

File: autobox.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+  if (!recps) return content
+
+  if (typeof recps === 'string') recps = content.recps = [recps]
+  if (!Array.isArray(recps)) throw new Error('private message field "recps" expects an Array of recipients')
+  if (recps.length === 0) throw new Error('private message field "recps" requires at least one recipient')
+
+  var ciphertext
+  for (var i = 0; i < boxers.length; i++) {
+    const boxer = boxers[i]
+    ciphertext = boxer(content, recps)
```
[the remark this answers is not in the document]
Comment (christianbundy) [reply autobox.js:17]: Since we're doing a breaking change anyway, it'd be cool to drop this. Discussed with Mix. He told me to type this.

[the remark this answers is not in the document]
Comment (mixmix) [reply autobox.js]: DONE

**Comment by mixmix** [comment]
@christianbundy the changes we talked about have been added.
I'm merging this into master (but we'll delay publishing a major version till we get those close tests passing)
